lib/protobuf/json.ex

```diff
@@ -0,0 +1,99 @@
+  | `message`                | `object`              | Yes       |
+  | `enum`                   | `string`              | Yes       |
+  | `map<K,V>`               | `object`              | Yes       |
+  | `repeated V`             | `array [v, …]`        | Yes       |
+  | `bool`                   | `true, false`         | Yes       |
+  | `string`                 | `string`              | Yes       |
+  | `bytes`                  | `base64 string`       | Yes       |
+  | `int32, fixed32, uint32` | `number`              | Yes       |
+  | `int64, fixed64, uint64` | `string`              | Yes       |
+  | `float, double`          | `number`              | Yes       |
+  | `Any`                    | `object`              | No        |
+  | `Timestamp`              | `string`              | No        |
+  | `Duration`               | `string`              | No        |
+  | `Struct`                 | `object`              | No        |
+  | `Wrapper types`          | `various types`       | No        |
+  | `FieldMask`              | `string`              | No        |
+  | `ListValue`              | `array [foo, bar, …]` | No        |
+  | `Value`                  | `value`               | No        |
+  | `NullValue`              | `null`                | No        |
+  | `Empty`                  | `object`              | No        |
+
+  ## Usage
+
+  `Protobuf.JSON` requires a JSON library to work, so first make sure you have `:jason` added
+  to your dependencies:
+
+      defp deps do
+        [
+          {:jason, "~> 1.2"},
+          # ...
+        ]
+      end
+
+  With `encode/1` you can turn any `Protobuf` message struct into a JSON string:
+
+      iex> message = Car.new(color: :RED, top_speed: 125.3)
+      iex> Protobuf.JSON.encode(message)
+      {:ok, "{\\"color\\":\\"RED\\",\\"topSpeed\\":125.3}"}
+
+  JSON keys are encoded as camelCase strings, specified by the `json_name` field options. So
+  make sure to recompile the `.proto` files in your project before working with JSON encoding.
+  """
+
+  alias Protobuf.JSON.{Encode, EncodeError}
+
+  @type encode_opt :: {:json_library, module} | Encode.encode_opt()
+
+  @spec encode!(struct, [encode_opt]) :: String.t() | no_return
+  def encode!(struct, opts) do
+    case encode(struct, opts) do
+      {:ok, json} -> json
+      {:error, error} -> raise error
+    end
+  end
+
+  @spec encode(struct, [encode_opt]) ::
+          {:ok, String.t()} | {:error, EncodeError.t() | Exception.t()}
+  def encode(struct, opts \\ []) do
+    with {:ok, props} <- get_props(struct),
+         {:ok, json_library} <- get_json_library(opts) do
+      struct
+      |> Encode.encode(props, opts)
+      |> json_library.encode()
+    end
+  end
+
+  defp get_props(%module{}) do
+    case module.__message_props__() do
+      %{syntax: :proto3} = props ->
+        {:ok, props}
+
+      %{syntax: syntax} ->
+        {:error, EncodeError.new({:unsupported_syntax, syntax})}
```

Comment:
Not sure if it is possible to embed proto2 messages into proto3 and vice versa. If it is, we should probably move this validation in, so we check it recursively too.

Reply:
Just found out [it is possible](https://developers.google.com/protocol-buffers/docs/proto#using_proto3_message_types). I think the safest choice would be to move this check inside. Will do it.